### ISSUE TITLE
pkcs#12

### ISSUE THREAD

**Issue by fkooman** [eduvpn/android#75]
In API 2.0 we have the option to separately get a certificate/key for the user. It would be nice to also be able to import those in the Android keychain. 

On iOS we the profile still needs to contain the CA, on Android not it seems and it can be part of the imported PKCK#12.

**Comment by fkooman** [comment]
@dzolnai is this possible?

**Comment by dzolnai** [comment]
I'm not a keychain expert, but it should probably work.
But it does not have an advantage in security, since the VPN profile contains the same cert / key, which is also saved on the permanent storage by the library.

**Comment by fkooman** [comment]
That's probably not true. With API 2.0 you download a profile_config separate from the cert/key (different API call) so there is no need to store the cert/key in "permanent" storage anymore if it is placed in the keychain. Or do I miss something?

**Comment by dzolnai** [comment]
What I'm saying is, that those properties are injected into the final profile used by the lib, which is then saved on the storage.

**Comment by fkooman** [comment]
That's not how the keychain works I hope! Once you import the private key it will never leave the secure storage anymore, it will only be used through the keychain. There is probably a special configuration flag for triggering this. The OpenVPN for Android app exposes this in the UI, so you can see how they do it there...

And what storage are you referring to? The private application storage only available to the eduVPN app I assume?

**Comment by dzolnai** [comment]
> The OpenVPN for Android app exposes this in the UI, so you can see how they do it there

I checked the sources, it is indeed supported, but not the default behavior.
> That's not how the keychain works I hope

I was talking about the VPN profile used by the library, not about the keychain. The library serializes the contents of the profile so that it can be reused later.
> And what storage are you referring to?

SD card or internal storage
> The private application storage only available to the eduVPN app I assume?

Yes, each application has it's own storage space under its own UID. But if malicious code can elevate to root permissions, it can read the files of all other apps.

To sum up, since I've found code for it in the library, it should be possible to do. Thanks for the pointer.

**Comment by fkooman** [comment]
> I was talking about the VPN profile used by the library, not about the keychain. The library serializes the contents of the profile so that it can be reused later.

Which library? The one you wrote? Or the OpenVPN for Android "as a library"? 

> SD card or internal storage

Your store private key and certificate on the SD card/internal storage available to all apps? Can this be changed? Sounds like a very bad idea!

> Yes, each application has it's own storage space under its own UID. But if malicious code can elevate to root permissions, it can read the files of all other apps.

But this is a lot less bad than allowing any application to read the VPN profile and send it somewhere :) We assume your phone is not "rooted". The keychain in addition provides protection against exactly this situation: when the phone is "rooted" the key can still not be extracted.

**Comment by dzolnai** [comment]
> Which library? The one you wrote? Or the OpenVPN for Android "as a library"?
OpenVPN for Android
> Your store private key and certificate on the SD card/internal storage available to all apps?
It is on the SD card / internal storage, but not available to all apps, only to EduVPN. I think we misunderstood each other.
> The keychain in addition provides protection against exactly this situation: when the phone is "rooted" the key can still not be extracted.
Sure, what I meant was this: since the OpenVPN for Android library stores the VPN profile data in the own application storage space, it is not any more secure, because the data is duplicated on a less-secure storage device. 
But since then you discovered that it has support for storing data in the keychain, so probably it stores a partial version without those keys on the storage only, so yes, I think this is possible.

**Comment by fkooman** [comment]
> It is on the SD card / internal storage, but not available to all apps, only to EduVPN. I think we misunderstood each other.

https://developer.android.com/guide/topics/data/data-storage.html  which one of these are we talking about? :) 

> But since then you discovered that it has support for storing data in the keychain, so probably it stores a partial version without those keys on the storage only, so yes, I think this is possible.

Right. Do you have any idea if this is difficult to implement? Probably something for later though! As long as other apps cannot obtain the private key of the VPN configuration it is not so bad :)

**Comment by dzolnai** [comment]
> which one of these are we talking about? :)
Shared Preferences, which is a key-value storage mapped to an XML file on the internal storage. The file is created with the [MODE_PRIVATE](https://developer.android.com/reference/android/content/Context.html#MODE_PRIVATE) flag.

> Do you have any idea if this is difficult to implement?
Depends a bit on how much of this functionality is exposed from the OpenVPN library to implementing apps. I think it shouldn't be too hard.

**Comment by fkooman** [comment]
When I was playing with this, around the time Android Studio took less than 10 minutes to only start on my laptop, the difficulty was in _importing_ the key/cert in the keychain. I think it only accepted pkcs#12 and not just a simple cert+key. But then again, my Android knowledge was/is pretty minimal :)

**Comment by dzolnai** [comment]
Hopefully OpenVPN for Android already solved this ;)

**Comment by fkooman** [comment]
That's the thing, I don't think so :(

**Comment by fkooman** [comment]
Maybe someday in the future.

**Comment by fkooman** [comment]
Seems this is implemented now in OpenVPN for Android: https://github.com/schwabe/ics-openvpn/commit/b9ac2b15eac3e5e5f9dc89c948ec8278e2e7c1f9

**Comment by fkooman** [comment]
Hm, actually, it seems it was implemented a long time ago already.

**Comment by fkooman** [comment]
I guess it doesn't really matter where things are stored, as long as it is secure (i.e. no other apps can access it).
